pyop2/compilation.py

```diff
@@ -481,24 +537,33 @@ def __init__(self, code, argtypes):
     else:
         raise ValueError("Don't know how to compile code of type %r" % type(jitmodule))
 
-    platform = sys.platform
-    cpp = extension == "cpp"
-    if not compiler:
-        compiler = configuration["compiler"]
-    if platform.find('linux') == 0:
-        if compiler == 'icc':
-            compiler = LinuxIntelCompiler(cppargs, ldargs, cpp=cpp, comm=comm)
-        elif compiler == 'gcc':
-            compiler = LinuxCompiler(cppargs, ldargs, cpp=cpp, comm=comm)
+    cpp = (extension == "cpp")
+
+    # Sniff compiler from executable
+    exe = (configuration["cxx"] if cpp else configuration["cc"])
+    exe = exe or ("g++" if cpp else "gcc")
```

Comment:
```py
if cpp:
    exe = configuration["cxx"] or "g++"
else:
    exe = configuration["cc"] or "gcc"
```
?
Also the logic seems a little bit tangled. We have an `exe` variable that will default to `"gcc"` that then gets transformed to `"GNU"`, `"Cray"` etc and *then* gets converted to `LinuxGnuCompiler`, `LinuxCrayCompiler` etc.Can this not be streamlined to `configuration["cc"] -> Linux{Gnu,Cray}Compiler`?

Reply:
I think this is the right thing to do because `PYOP2_CC` would normally be set to something like `/opt/mpich/bin/mpicc` and there's not a direct route from this to the right compiler class without doing something like what `sniff_compiler_version` does.

The Cray compiler and ScoreP compilers spring to mind as examples of compilers that require some sniffing to work out what they really are.

This does raise the question of whether `gcc` is the correct default here though.